File: framework/languages/java/zlink-stream-connector/src/main/java/systems/zlink/stream/connector/ZLinkStreamAssert.java
package systems.zlink.stream.connector;

import java.io.IOException;
import java.util.Objects;
import java.util.concurrent.CompletionException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

public final class ZLinkStreamAssert {
    private ZLinkStreamAssert() {
    }

    public static void ensure(boolean condition, String message) {
        if (message == null || message.isBlank()) {
            throw new IllegalArgumentException("message is required");
        }
        if (!condition) {
            throw new IllegalStateException(message);
        }
    }

    public static ZLinkStreamError expectFailure(
        ThrowingRunnable action,
        String errorKind) {
        Objects.requireNonNull(action, "action");
        Throwable failure;
        try {
            action.run();
        } catch (Throwable error) {
            failure = unwrap(error);
            return requireKind(classify(failure), errorKind);
        }
        throw new IllegalStateException("Expected action to fail.");
    }

    private static ZLinkStreamError requireKind(
        ZLinkStreamError streamError,
        String errorKind) {
        if (errorKind != null && !streamError.code().name().equals(errorKind)) {
            throw new IllegalStateException(
                "Expected failure kind '" + errorKind + "', got '"
                    + streamError.code().name() + "'.",
                streamError.exception());
        }
        return streamError;
    }

    public static void expectTimeout(ThrowingRunnable action) {
        ZLinkStreamError error = expectFailure(action, null);
        if (error.code() != ZLinkStreamErrorCode.REQUEST_TIMEOUT
            && error.code() != ZLinkStreamErrorCode.CONNECT_TIMEOUT) {
            rethrow(error.exception());
        }
    }

    private static ZLinkStreamError classify(Throwable failure) {
        ZLinkStreamErrorCode code;
        if (failure instanceof TimeoutException) {
            code = failure.getMessage() != null && failure.getMessage().startsWith("connect timed out")
                ? ZLinkStreamErrorCode.CONNECT_TIMEOUT
                : ZLinkStreamErrorCode.REQUEST_TIMEOUT;
        } else if (failure instanceof IllegalArgumentException) {
            code = ZLinkStreamErrorCode.VALIDATION_FAILED;
        } else if (failure instanceof IOException) {
            code = ZLinkStreamErrorCode.DISCONNECTED;
        } else {
            code = ZLinkStreamErrorCode.REMOTE_ERROR;
        }
        String message = failure.getMessage();
        if (message == null || message.isBlank()) {
            message = failure.getClass().getSimpleName();
        }
        return new ZLinkStreamError(code, message, failure);
    }

    private static Throwable unwrap(Throwable error) {
        Throwable current = error;
        while ((current instanceof CompletionException || current instanceof ExecutionException)
            && current.getCause() != null) {
            current = current.getCause();
        }
        return current;
    }

    private static void rethrow(Throwable error) {
        if (error instanceof RuntimeException runtime) {
            throw runtime;
        }
        if (error instanceof Error fatal) {
            throw fatal;
        }
        throw new CompletionException(error);
    }

    @FunctionalInterface
    public interface ThrowingRunnable {
        void run() throws Exception;
    }
}

File: framework/languages/dotnet/src/Systems.Zlink.Stream.Connector/Contracts/ZlinkStreamAssert.cs
namespace Systems.Zlink.Stream.Connector.Contracts;

using System.Runtime.ExceptionServices;

/// <summary>
///     Provides assertions that do not depend on connector state.
/// </summary>
public static class ZlinkStreamAssert
{
    /// <summary>
    ///     Throws with the required diagnostic message when the condition is false.
    /// </summary>
    public static void Ensure(bool condition, string message)
    {
        if (string.IsNullOrWhiteSpace(message))
            throw new ArgumentException("Assertion message is required.", nameof(message));
        if (!condition) throw new InvalidOperationException(message);
    }

    /// <summary>
    ///     Executes an action, requires it to fail, and optionally verifies its error kind.
    /// </summary>
    public static async ValueTask<ZlinkStreamError> ExpectFailureAsync(
        Func<CancellationToken, ValueTask> action,
        string? errorKind = null)
    {
        ArgumentNullException.ThrowIfNull(action);

        try
        {
            await action(CancellationToken.None).ConfigureAwait(false);
        }
        catch (Exception exception)
        {
            var error = Classify(exception);
            if (errorKind is not null
                && !string.Equals(error.Code.ToString(), errorKind, StringComparison.Ordinal))
                throw new InvalidOperationException(
                    $"Expected failure kind '{errorKind}', got '{error.Code}'.",
                    exception);
            return error;
        }

        throw new InvalidOperationException("Expected action to fail.");
    }

    /// <summary>
    ///     Executes an action and requires a timeout failure; other failures are propagated.
    /// </summary>
    public static async ValueTask ExpectTimeoutAsync(
        Func<CancellationToken, ValueTask> action)
    {
        ArgumentNullException.ThrowIfNull(action);

        try
        {
            await action(CancellationToken.None).ConfigureAwait(false);
        }
        catch (Exception exception)
        {
            if (TryClassifyKnownFailure(exception, out var error)
                && error.Code is ZlinkStreamErrorCode.RequestTimeout or ZlinkStreamErrorCode.ConnectTimeout)
                return;

            ExceptionDispatchInfo.Capture(exception).Throw();
        }

        throw new InvalidOperationException("Expected action to time out.");
    }

    private static ZlinkStreamError Classify(Exception exception)
    {
        if (TryClassifyKnownFailure(exception, out var error)) return error;

        ExceptionDispatchInfo.Capture(exception).Throw();
        throw new InvalidOperationException("Unreachable code.");
    }

    private static bool TryClassifyKnownFailure(
        Exception exception,
        out ZlinkStreamError error)
    {
        if (exception is ZlinkStreamException streamException)
        {
            error = streamException.Error;
            return true;
        }

        for (var current = exception; current is not null; current = current.InnerException)
        {
            var code = current switch
            {
                TimeoutException => ZlinkStreamErrorCode.RequestTimeout,
                ArgumentException => ZlinkStreamErrorCode.ValidationFailed,
                HttpRequestException => ZlinkStreamErrorCode.Disconnected,
                IOException => ZlinkStreamErrorCode.Disconnected,
                _ => (ZlinkStreamErrorCode?)null
            };
            if (code is { } knownCode)
            {
                error = new ZlinkStreamError(knownCode, exception.Message, exception);
                return true;
            }
        }

        error = null!;
        return false;
    }
}
